internal/gitserver/stream_hunks.go

```diff
@@ -0,0 +1,234 @@
+package gitserver
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// blameHunkReader enables to read hunks from an io.Reader.
+type blameHunkReader struct {
+	hunks chan hunkResult
+}
+
+func newBlameHunkReader(ctx context.Context, rc io.ReadCloser) HunkReader {
+	br := &blameHunkReader{
+		hunks: make(chan hunkResult),
+	}
+	go br.readHunks(ctx, rc)
+	return br
+}
+
+type hunkResult struct {
+	hunk *Hunk
+	err  error
+}
+
+func (br *blameHunkReader) readHunks(ctx context.Context, rc io.ReadCloser) {
+	newHunkParser(rc, br.hunks).parse(ctx)
+}
+
+// Read returns a slice of hunks, along with a done boolean indicating if there is more to
+// read.
+func (br *blameHunkReader) Read() ([]*Hunk, bool, error) {
+	res, ok := <-br.hunks
+	if !ok {
+		return nil, true, nil
+	}
+	if res.err != nil {
+		return nil, false, res.err
+	} else {
+		return []*Hunk{res.hunk}, false, nil
+	}
+}
+
+type hunkParser struct {
+	rc      io.ReadCloser
+	sc      *bufio.Scanner
+	hunksCh chan hunkResult
+
+	// commits stores previously seen commits, so new hunks
+	// whose annotations are abbreviated by git can still be
+	// filled by the correct data even if the hunk entry doesn't
+	// repeat them.
+	commits map[string]*Hunk
+}
+
+func newHunkParser(rc io.ReadCloser, hunksCh chan hunkResult) hunkParser {
+	return hunkParser{
+		rc:      rc,
+		hunksCh: hunksCh,
+
+		sc:      bufio.NewScanner(rc),
+		commits: make(map[string]*Hunk),
+	}
+}
+
+func (p hunkParser) parse(ctx context.Context) {
+	defer p.rc.Close()
+	defer close(p.hunksCh)
+
+	var cur *Hunk
+	for {
+		if err := ctx.Err(); err != nil {
+			return
+		}
+
+		// Do we have more to read?
+		if !p.sc.Scan() {
+			if cur != nil {
+				if h, ok := p.commits[string(cur.CommitID)]; ok {
+					cur.CommitID = h.CommitID
+					cur.Author = h.Author
+					cur.Message = h.Message
+				}
+				// If we have an ongoing entry, send it.
+				p.hunksCh <- hunkResult{hunk: cur}
```

Comment:
```suggestion
				select {
					case <- p.hunksCh <- hunkResult{hunk: cur}:
					case <- ctx.Done():
						return
				}
```

I think every write needs this as you're not draining the channel (on the consumer side) to ensure that all writes succeed when the context is canceled and the consumer bails out.

Reply:
Thank you 🙏